### ISSUE TITLE
Should we rewrite the library in JavaScript?

### ISSUE THREAD

**Issue by yellowiscool** [SINTEF-9012/PruneCluster#79]
The library is written in [TypeScript](http://www.typescriptlang.org/). TypeScript is a superset of JavaScript adding strong typing and other features such as classes and interfaces.

When I started the development, TypeScript was brand new and promising. 1 year later, it is not very popular in the JavaScript community. I think ES6 with tools such as [Babel](http://babeljs.io/) have reduced the interest in TypeScript. It still has types, but bug reports demonstrated that the library needs to check everything during runtime anyway.

I personally like TypeScript and I think the code completion and type checking features are pretty neat when used inside VisualStudio with [Leaflet's DefinitelyTyped definitions](http://definitelytyped.org/). But even though it's very light and close to ES6's syntax, it's a different language with a low popularity.

The conversion from TypeScript to ES6 (with Babel or equivalent for backward compatibility) can be considered.

**Comment by itbeyond** [comment]
I think the library is excellent and I for one would be happy to add extensions and update like my new Layer Switch that applies filters to prunecluster. I have never even looked at TypeScript so have no idea how hard a conversion, the source it creates in JS is not unreadable. I will be using this library and think it is well worth moving forward.

**Comment by FalloutPL** [comment]
I think that TS have more pros than cons so i am against it. It is well maintained and gives many more every new release. Personally i cant think about writing again in vanilla js :-) His time to shine is about to start when Angular2 become to be widely used.

**Comment by dtruel** [comment]
I think it would be worthwhile - it makes it easier to make changes.  That in turn makes it easier to do pull requests with modifications I have made in plain JS.

**Comment by w8r** [comment]
First of all - great library, superb performance!
We can do it using ES2015 + some type checks. Anyway, not to offend anyone, the `C#` code style is very confusing and not compatible with Leaflet core code.

**Comment by dtruel** [comment]
Just a note: I believe that AngularJS 2 is being written in TypeScript.  So I don't know what that means for typescript, but it definitely is a plus.

**Comment by yellowiscool** [comment]
Yes Angular 2 is written in TypeScript and the language [looks more and more popular](https://www.google.com/trends/explore#q=typescript). However it's still a small community compared to JavaScript.

A solution might be rewriting the code in JavaScript using Leaflet coding style this time, but also maintaining a TypeScript definitions file.

**Comment by kasparsklavins** [comment]
I support the idea of rewriting it to ES2015.
Typescript releases are far to common for IDE's to keep up.
